### DISCUSSION TITLE
Ideas concering usability

### DISCUSSION THREAD

**Discussion by ghost** [YuvrajRaghuvanshiS/WhatsApp-Key-Database-Extractor#10] [General]
Hello,
so during our conversation, I got further than before.

Just some things I want to mention:
1. Sorry for the critique!
2. I started the whole thing with "python wa_kdbe.py 0 0" , now there was an error that the folder extracted/0 doesn't exist.
3.  Also maybe mention that one has to copy paste the device name when prompted (possibility to automate it if only one device is connected?)
Anyway, thanks alot for your work!

**Comment by YuvrajRaghuvanshiS** [comment]
starting from point 3 I am currently working on that. Also I understand while geeking out for this project I unintentionally made this overwhelming.
Try with alphabets. Like python wa_kdbe.py qwer rschuppan

**Comment by ghost** [comment]
Another point maybe: (4) I dont think the user function /parameter is really necessary (functionality wise), it just adds a sprinkle of confusion.

**Reply by YuvrajRaghuvanshiS** [reply]
Yes, sarcasm, like I'm not very full of it already. Or maybe perfectionism is really bad for my health. It was used just in case you own more than one device. But now I'm thinking of making user as "user" and add a condition to check if it already exists. But it has it's own complication. I need to work on that too.

**Reply by ghost** [reply]
If you leave default values it should be fine I think (so one can just go "python wa_kdbe.py"), nothing toooo big a problem.

**Reply by YuvrajRaghuvanshiS** [reply]
You mean removing command line parameters? Like this? https://github.com/YuvrajRaghuvanshiS/WhatsApp-Key-Database-Extractor/tree/4577a638a2af903e472e56bd80bc2bcd8a55a96f

**Reply by ghost** [reply]
yes! I mean, I couldn't really figure out what the password is for, and we established that the user is not necessary

**Reply by YuvrajRaghuvanshiS** [reply]
I am going to revert this entire overwhelming and confusion causing process to what it used to be earlier.

**Reply by ghost** [reply]
While your at it, PLEASE correct this spelling mistake: "Run wa-kdbe.py    python3 wa-kdbe.py" 
Your file is called "**wa_kdbe.py**" XD ;-)

**Reply by YuvrajRaghuvanshiS** [reply]
ohh, sure.

**Reply by ghost** [reply]
maybe also shorten the "1." point (clone the git).
Just put "1.  Clone (download) the tool" there, improves readability

**Reply by YuvrajRaghuvanshiS** [reply]
#ToDo

**Reply by ghost** [reply]
should I just write a revised version of readme and put it in apull request?

**Reply by YuvrajRaghuvanshiS** [reply]
That is quick fix, I can do it.

**Comment by ghost** [comment]
"If you do not see any errors in above lines in extracting/fluffing whatsapp.ab you SHOULD choose to clean temporary folder. It contains your chats in UN-ENCRYPTED format."
ARE YOU SERIOUS? Omg thats awesome, which file is it?

**Reply by ghost** [reply]
the chats I mean, so I can use it for https://community.signalusers.org/t/signal-fork-with-whatsapp-migration

**Reply by YuvrajRaghuvanshiS** [reply]
It is whatsapp.tar file after converting from whatsapp.ab,  Yes you can use it but let me do it. It will decompress and move those important files to extracted/<userName> folder.

**Reply by ghost** [reply]
what do you mean with "let you do it"? Is it not enough if I just rename the file?

**Reply by YuvrajRaghuvanshiS** [reply]
I meant let this tool do it.
After "If you do not see any errors in above lines in extracting/fluffing whatsapp.ab you SHOULD choose to clean temporary folder. It contains your chats in UN-ENCRYPTED format." If there's no error you can delete those files it's really upto you however msgstore.db is already in extracted folder by this time

**Reply by ghost** [reply]
Oh. my. god. You're the best! my dumb ass was searching in tmp XD
(msgstore.db is unencrypted right?)

**Reply by YuvrajRaghuvanshiS** [reply]
yes it is unencrypted.

**Reply by ghost** [reply]
Okay this has to be promoted WAAAAAAY more

**Reply by YuvrajRaghuvanshiS** [reply]
#ToDo. LoL

**Reply by ghost** [reply]
I searched for a whole friggin week for
1. A way to get the database (yes I can just copy it I know dont laugh)
2. A way to get the key
3. A way to decrypt the database.
Yours is currently the ONLY TOOL that fullfills both 2 & 3!

**Reply by YuvrajRaghuvanshiS** [reply]
I was not laughing. Internet is overwhelming. And my god I using this word so much that I feel like that "Apparently Kid"

**Reply by ghost** [reply]
Proposing to rename your tool to WA_Chat_Extractor.
I mean what else would one do with a key? get the chats. So thats the ultimate goal and I think that should be the focus on.

**Reply by YuvrajRaghuvanshiS** [reply]
Key can decrypt msgstore.db.crpyt12 file which is easily accessible.

**Reply by ghost** [reply]
Yes, thats true, but your tool already delivers a decrypted version. And AFAIK there is no working decryptor out there right now.

**Reply by ghost** [reply]
My point is: The only reason one needs this god forsaken key (I really hate it) is to get their msgstore.db decrypted. 
But if one already gets a decryted copy WITH THE SAME TOOL, one doesn't need the key anymore.

**Reply by YuvrajRaghuvanshiS** [reply]
Actually there are many decryptors. And some do work with termux (terminal emulator for android)

**Reply by ghost** [reply]
Yes, but I couldn't find one (excluding Termux) that works with current Android & WhatsApp.

**Reply by YuvrajRaghuvanshiS** [reply]
This title increases search indexing on github.

**Comment by YuvrajRaghuvanshiS** [comment]
=== Likely ToDo ===
1. Remove CL arguments.
2. Do not check for more than one device.
3. Use default username.

**Reply by ghost** [reply]
I would edit 2: Only prompt for device name if more than one device is found by adb. Should be a simple if-clause, right?

**Reply by YuvrajRaghuvanshiS** [reply]
yes but inside that "if" I need to check if there are more that two devices there's no single way of doing that and regex is different for windows based and linux based. It is completely doable but will take time which I lack.

**Reply by ghost** [reply]
okay yeah I understand, then just put "copy your devices name from the device list if prompted" in the readme

**Reply by YuvrajRaghuvanshiS** [reply]
Better just add "Remove all other device and turn off android emulator."

**Reply by ghost** [reply]
True!

**Reply by YuvrajRaghuvanshiS** [reply]
#TODO point 2, No need of copy pasting device id, just select from list provided.  5e66774

`List of devices attached`

`1. 01234ab5c67 device  Redmi 5 Plus`
`2. ab0123c4 device  Redmi Note 9 Pro`
`Enter device number (for ex : 2) :`

Working on skipping entirely if only one device is attached.

**Reply by YuvrajRaghuvanshiS** [reply]
#TODO point 2 done : Do not ask for device if only one is connected in 4af4bed

**Reply by ghost** [reply]
Great work man👍

**Comment by ghost** [comment]
If you want, I'd love to contribute!

**Reply by YuvrajRaghuvanshiS** [reply]
https://github.com/YuvrajRaghuvanshiS/WhatsApp-Key-Database-Extractor/tree/revert/argparse
This reverts command line arguments. Point 1. of likely ToDo. I know this is okay but if want you can test it.

git clone https://github.com/YuvrajRaghuvanshiS/WhatsApp-Key-Database-Extractor.git
cd WhatsApp-Key-Database-Extractor
git checkout revert/argparse
python wa_kdbe.py

**Reply by ghost** [reply]
will do

**Reply by YuvrajRaghuvanshiS** [reply]
No it should not. Can you send me a screenshot where it does? And one of terminal `git status`.

**Reply by ghost** [reply]
used the wrong version. But there's a different problem now:
![image](https://user-images.githubusercontent.com/31773450/106022218-e44ce300-60c5-11eb-83a9-9d86b957e48e.png)

**Reply by ghost** [reply]
I remember, this was why I didn't use this tool the first time

**Reply by ghost** [reply]
> used the wrong version. But there's a different problem now:
> ![image](https://user-images.githubusercontent.com/31773450/106022218-e44ce300-60c5-11eb-83a9-9d86b957e48e.png)

okay so this happens if one doesn't run "adb devices" and confirms on phone first. Definitly something to put in troubleshooting

**Reply by YuvrajRaghuvanshiS** [reply]
> 
> 
> used the wrong version. But there's a different problem now:
> ![image](https://user-images.githubusercontent.com/31773450/106022218-e44ce300-60c5-11eb-83a9-9d86b957e48e.png)

#ToDo adb kill_server before showing devices.
I need to push a commit but can edit ADBDeviceSerialId.py add at line 19 `os.system(adb + ' kill-server')`

**Reply by ghost** [reply]
Correction: running "adb devices" doesn't fix the (I now call it) empty device list bug.

**Reply by ghost** [reply]
![image](https://user-images.githubusercontent.com/31773450/106022923-9d132200-60c6-11eb-8f91-5b5ba934701d.png)
really interesting, do you know what this means?

**Reply by ghost** [reply]
okay, so if I copy the device name from "adb devices" into the prompt, I get this error (the device list your tool showed was empty in this case):
![image](https://user-images.githubusercontent.com/31773450/106023161-d6e42880-60c6-11eb-85cc-936bdd0db839.png)

**Reply by YuvrajRaghuvanshiS** [reply]
> do you know what this

Yes. When you connect a phone with debugging for the first time it needs permission from user to be accepted on phone to allow debugging. This is that. "Allow debugging.... computer fingerprint is..."

**Reply by ghost** [reply]
what confuses me is, I did that

**Reply by YuvrajRaghuvanshiS** [reply]
> 
> 
> okay, so if I copy the device name from "adb devices" into the prompt, I get this error (the device list your tool showed was empty in this case):
> ![image](https://user-images.githubusercontent.com/31773450/106023161-d6e42880-60c6-11eb-85cc-936bdd0db839.png)

It was empty because it could not find any device. It allowed you in beginning but when it needed that device it was not there.

**Reply by ghost** [reply]
right now the only bypass seems to be the "always allow this computer" checkbox

**Reply by YuvrajRaghuvanshiS** [reply]
> 
> 
> what confuses me is, I did that

Tick that checkbox "Always allow"

**Reply by ghost** [reply]
yes, but it should work without that box, because it poses a security risk

**Reply by YuvrajRaghuvanshiS** [reply]
It does not as long as it is your computer and you turn off USB debugging in develper options later. Also below that toggle (Turn off USB debugging) there's "Revoke USB debugging authorizations" you can do that if you want.

**Reply by ghost** [reply]
yes. 
Concerning the bug, the tool should trigger an authorization prompt on the phone, so the list isn't empty: 
![image](https://user-images.githubusercontent.com/31773450/106024218-d8fab700-60c7-11eb-9536-9c4320a58661.png)

**Reply by ghost** [reply]
okay very interesting: 
Just did the variant with the "always allow" checkbox checked, and it still doesn't show up in the device list

**Reply by YuvrajRaghuvanshiS** [reply]
> 
> 
> okay very interesting:
> Just did the variant with the "always allow" checkbox checked, and it still doesn't show up in the device list

I am working on that. Removing This all enter device. ToDo point 2. I will just use `adb wait-for-devices` and this will be enough as long as user doesn't remove it

**Reply by ghost** [reply]
> > okay very interesting:
> > Just did the variant with the "always allow" checkbox checked, and it still doesn't show up in the device list
> 
> I am working on that. Removing This all enter device. ToDo point 2. I will just use `adb wait-for-devices` and this will be enough as long as user doesn't remove it

Yeah I think thats a good idea. Just produces confusion and errors.

**Reply by YuvrajRaghuvanshiS** [reply]
> 
> 
> okay very interesting:
> Just did the variant with the "always allow" checkbox checked, and it still doesn't show up in the device list

Fix it in 7e55a50

**Comment by ghost** [comment]
What's the latest on Termux by the way?

**Reply by YuvrajRaghuvanshiS** [reply]
Not really doing much with that branch. Can't afford time.

**Reply by ghost** [reply]
oh okay.
I thought we fixed it and found the problem.
